### PR TITLE
[WIP] Use system path and path extension vars when searching for maven binaries

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

In #1149, we implemented functionality to allow users to specify custom binaries for Maven analysis. At the same time we introduced rudimentary extensions for Maven binary searching: we first searched the project directory, then any of its ancestors, then any directories in the system path, for `mvn` or `mvnw`. In Windows, we'd search for these with the extensions `.exe` and `.bat`.

This PR significantly expands that functionality in Windows. As of this PR, we first search the project directory, then any of its ancestors, then any directories in the system path, for `mvn` or `mvnw`. In Windows, we search for these with any extension that Windows reports it can execute via `%PATHEXT%`.

We also stopped running `mvn -v` to determine if a command is viable; instead we now just check for the commands existence on disk and then try to use it if found. Similarly we no longer perform any validation on the binary passed via `$FOSSA_MAVEN_CMD`, other than validating that it exists on disk.

## Acceptance criteria

Windows users should experience significantly fewer bugs when running Maven analysis in Windows. If this works well, we can then extend this functionality to other command execution scenarios.

## Testing plan

Automated tests should cover this functionality.
As well, I ran this locally and validated that it looks in the correct spots via `ProcMon`. 

Example:
Screenshot 2023-05-25 123354

## Risks

This is yet more messing with how Maven binaries are searched and executed, leading to more potential confusion.

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [ ] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

### CODE REVIEW

File: src/Effect/Exec.hs [hunk 1]
```diff
@@ -287,6 +285,238 @@ instance ToDiagnostic ExecErr where
...
+-- The function can be provided with multiple binary names.
+-- This is useful in cases where a binary might have different names,
+-- but the usage of the binary is exactly the same regardless of the name used.
+-- The binary names are searched in order, and the function returns the path of the first one found.
+--
+-- * Examples:
+--
+-- __Windows__
+--
+-- > Binaries: [ "mvn", "mvnw" ]
```
Comment (spatten) [inline src/Effect/Exec.hs:399]: If, in practice, mvnw is preferred over mvn, then I think this example should reflect this

File: src/Strategy/Maven/Plugin.hs [hunk 2]
```diff
@@ -116,20 +117,23 @@ withUnpackedPlugin plugin act =
 
       act pluginJarFilepath
 
-installPlugin :: (CandidateCommandEffs sig m, Has ReadFS sig m) => Path Abs Dir -> FP.FilePath -> DepGraphPlugin -> m ()
+type MavenEffs sig m = (Has Exec sig m, OverrideWhichEffs sig m)
```
Comment (spatten) [inline src/Strategy/Maven/Plugin.hs:120]: The `MagenEffs` type really cleans things up!

**Review by spatten — APPROVED** [review]
A few minor comments, but other than that it looks great

File: src/Effect/Exec.hs (hunk 1 above)
Comment (jssblck) [reply src/Effect/Exec.hs]: So uh, it turns out, that I had messed up the actual order 👀 
But it's [fixed now in the latest commit](https://github.com/fossas/fossa-cli/pull/1209/commits/e187e49071c84254eb4cdea45f0568e93b4d65a5#diff-44189d9cbe979286ff7d5111b3e7f795518942972009ea8f436f855c8438b82e)...

Comment (jssblck) [reply src/Effect/Exec.hs]: Glad you noticed this!

File: docs/features/strategy-command-selection.md [hunk 3]
```diff
@@ -6,64 +6,113 @@ FOSSA refers to such strategies as "dynamic analysis strategies".
...
+PATHEXT : ".bat;.exe"
+SCANDIR : "C:\Users\me\projects\example"
 ```
 
-And that FOSSA CLI is being run with the following command:
+Searches:
+
 ```
-FOSSA_MAVEN_CMD=/usr/local/bin/custom-mvn fossa analyze
+C:\Users\me\projects\example\mvn.exe
```
[the remark this answers is not in the document]
Comment (jssblck) [reply docs/features/strategy-command-selection.md]: Fixed!

File: docs/features/strategy-command-selection.md [hunk 4]
```diff
@@ -6,64 +6,113 @@ FOSSA refers to such strategies as "dynamic analysis strategies".
...
 ```
 
-FOSSA CLI evaluates these candidates with the following commands in sequence,
-stopping after the first one evaluates with exit code zero (signifying no error):
+Searches:
+
 ```
-; /usr/local/bin/custom-mvn -v # exit code 1
-; ./mvnw -v # exit code 0
+/home/me/projects/example/mvn
```
[the remark this answers is not in the document]
Comment (jssblck) [reply docs/features/strategy-command-selection.md]: Fixed!

**Comment by jssblck** [comment]
We've run into a blocker here where this change alters existing functionality; putting in draft for now.

**Comment by jssblck** [comment]
Closing this to reduce open PRs, we can reopen as we decide to invest in this more.

**Pull request closed without merging**